### PR TITLE
Add Apothecary contract

### PR DESCRIPTION
**Pull request by Mathieu-Be**
**Apothecary contract:**

Each epoch, every doctor can attempt to brew a potion using `makePotion`. Based on a random number provided by Chainlink VRF, each `makePotion` will have a tiny chance to win a potion, that will be sent to the doctor owner.

Potions meant to be sent to winners need to be added to the contract using `addPotion`.

On the first attempt of a doctor, success will be guarantied. In this case, the potion will be pulled from the potion contract using `devMint`, as it will be a Launchpeg contract. This replaces the airdrop and allows people to claim their potions.

### CODE REVIEW

File: src/Apothecary.sol [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+        uint256 allLogsCount = allBrewLogs.length;
+
+        for (uint256 i = 0; i < allLogsCount; ++i) {
+            if (i == RECENT_BREW_LOGS_COUNT) {
+                break;
+            }
+
+            lastBrewLogs[i] = allBrewLogs[offsetIndex + i];
+        }
+    }
```
Comment (0x0Louis) [inline src/Apothecary.sol]: This function will return a bunch of zeros until there are enough logs.
The latest logs will be the last item which is a bit weird to me. Let's use this function that fixes both issues (and that can be easily adapted for pagination if needed):
```solidity
    function getlatestBrewLogs() external view override returns (BrewLog[] memory lastBrewLogs) {
        uint256 allLogsCount = allBrewLogs.length;
        uint256 length = allLogsCount > RECENT_BREW_LOGS_COUNT ? RECENT_BREW_LOGS_COUNT : allLogsCount;
        lastBrewLogs = new BrewLog[](length);

        for (uint256 i = 0; i < length; ++i) {
            lastBrewLogs[i] = allBrewLogs[allLogsCount - i - 1];
        }
    }
```

File: src/Apothecary.sol [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+        uint256 checkedLength = _count < totalDoctorBrews ? _count : totalDoctorBrews;
+        BrewLog[] memory lastNBrewLogs = new BrewLog[](checkedLength);
+
+        uint256 j = totalDoctorBrews;
+        for (uint256 i = checkedLength; i > 0;) {
+            lastNBrewLogs[i - 1] = _doctorBrewLogs[_doctorId][j - 1];
+            unchecked {
+                --i;
+                --j;
+            }
```
Comment (0x0Louis) [inline src/Apothecary.sol:184]: Moving the unchecked block a line before would avoid doing `i - 1` and `j - 1`, but as you prefer tbh

File: src/Apothecary.sol [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+        _doctorBrewLogs[_doctorId].push(brewLog);
+        allBrewLogs.push(brewLog);
+    }
+
+    /// @notice Returns period start of epoch timestamp
+    /// @param _epochTimestamp Timestamp of epoch
+    /// @return epochStart Start timestamp of epoch
+    function _getEpochStart(uint256 _epochTimestamp) private view returns (uint256 epochStart) {
+        uint256 startTimeCached = startTime;
+        epochStart = startTimeCached + ((_epochTimestamp - startTimeCached) / EPOCH_DURATION) * EPOCH_DURATION;
```
Comment (0x0Louis) [inline src/Apothecary.sol:385]: Reverts on `_epochTimestamp < startTimeCached`, but that might be wanted

File: src/Apothecary.sol [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+    function getVRFForEpoch(uint256 _epochTimestamp) external view override returns (uint256 epochVRF) {
+        epochVRF = _epochVRFNumber[_getEpochStart(_epochTimestamp)];
+    }
+
+    /// @notice Returns true if plague doctor attempted to brew a potion in an epoch
+    /// @notice and false otherwise
+    /// @param _epochTimestamp Timestamp of epoch
+    /// @param _doctorId Token ID of plague doctor
+    /// @return tried Boolean showing plague doctor brew attempt in epoch
+    function getTriedInEpoch(uint256 _epochTimestamp, uint256 _doctorId) external view override returns (bool tried) {
```
Comment (0x0Louis) [inline src/Apothecary.sol]: rename this to `triedToBrewPotionDuringEpoch`

File: src/Apothecary.sol [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+                uint256 pendingRequestId = _epochRequestId[nextEpochTimestampCache];
+
+                if (pendingRequestId != 0) {
+                    revert VrfRequestPending(pendingRequestId);
+                }
+
+                plagueDoctorVRFCaller = _doctorId;
+                _epochRequestId[nextEpochTimestampCache] = _vrfCoordinator.requestRandomWords(
+                    _keyHash, _subscriptionId, VRF_BLOCK_CONFIRMATIONS, _maxGas, RANDOM_NUMBERS_AMOUNT
+                );
```
Comment (0x0Louis) [inline src/Apothecary.sol:256]: if `makePotion` is first called at the end of an epoch, `fulfillRandomWords` may be called during the next epoch which will make the VRF reverts and `makePotion` would need to be called again.
I don't think this poses a DOS issue, but please write a test for this specific edge case in case it were to happen

File: src/Apothecary.sol [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+    /// @notice Returns first token ID of potions owned by Apothecary contract
+    /// @dev Reverts if no potions is owned by Apothecary contract
+    /// @return potionId First potion ID owned by Apothecary contract
+    function _getPotionId() private returns (uint256 potionId) {
+        if (_getPotionsLeft() == 0) {
+            revert PotionsNotEnough(0);
+        }
+        potionId = _potionsOwnedByContract[_potionsOwnedByContract.length - 1];
+        _potionsOwnedByContract.pop();
+    }
```
Comment (0x0Louis) [inline src/Apothecary.sol]: Scary to pop the array while not transferring it, please refactor this function to `sendPotion`

File: src/Apothecary.sol [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+                );
+            } else {
+                _brew(_doctorId);
+            }
+        } else {
+            potions.devMint(1);
+            hasMintedFirstPotion[_doctorId] = true;
+
+            potions.transferFrom(address(this), doctors.ownerOf(_doctorId), potions.totalSupply() - 1);
+        }
```
Comment (0x0Louis) [inline src/Apothecary.sol:265]: This should be a `if (!hasMintedFirstPotion[_doctorId])` as there are no reasons not to be able to brew a potion while receiving the first potion?

File: src/Apothecary.sol [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+
+    /// @notice Transfers potions from Apothecary contract to owner
+    /// @dev Potion IDs should be owned by Apothecary contract
+    /// @param _amount Number of potions to be transferred from Apothecary contract to owner
+    function removePotions(uint256 _amount) external override onlyOwner {
+        for (uint256 i = 0; i < _amount; ++i) {
+            potions.transferFrom(address(this), msg.sender, _getPotionId());
+        }
+        emit PotionsRemoved(_amount);
+    }
```
Comment (0x0Louis) [inline src/Apothecary.sol]: I don't get why we don't just mint the potion directly when brewing is successful.
If it's to have control on emission, just add a cap on the mint supply that can be increased by the owner (our multisig) while minting would be allowed to this contract only

File: src/Apothecary.sol [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+        if (plagueGame.isGameOver()) {
+            revert GameIsClosed();
+        }
+        ++totalBrewsCount;
+
+        BrewLog memory brewLog;
+        _triedBrewInEpoch[latestEpochTimestamp][_doctorId] = true;
+        bytes32 hash = keccak256(abi.encodePacked(_epochVRFNumber[latestEpochTimestamp], _doctorId));
+
+        if (uint256(hash) % difficulty == 0) {
```
Comment (0x0Louis) [inline src/Apothecary.sol:350]: I would do `uint256(hash) % PRECISION >= difficulty` so you can choose the difficulty percentage and not just `1/x` values (e.g, you can't do 60%, 40% etc...). If done that way, `difficulty` can be in the `[0, PRECISION[` range

**Review by 0x0Louis — CHANGES_REQUESTED** [review]

File: src/Apothecary.sol (hunk 3 above)
Comment (Mathieu-Be) [reply src/Apothecary.sol:385]: Yeah you don't need the  `epochStart` if the game hasn't started

File: src/Apothecary.sol (hunk 5 above)
Comment (Mathieu-Be) [reply src/Apothecary.sol:256]: Good point, added a test. Basically the first VRF callback would fail but with no bad consequence

File: src/Apothecary.sol (hunk 7 above)
Comment (Mathieu-Be) [reply src/Apothecary.sol:265]: I think it's cool if we only distribute at most 1 potion per call, it would kill the surprise if you receive a winning potion at the same time as claiming

File: src/Apothecary.sol (hunk 8 above)
Comment (Mathieu-Be) [reply src/Apothecary.sol]: Will be easier to just use an existing contract as Launchpeg

**Review by 0x0Louis — APPROVED** [review]
